### PR TITLE
Do no-type-slots dictoffset check as python lookup

### LINKED ISSUE #5696: Failing cdef_multiple_inheritance_nodict test when CYTHON_USE_TYPE_SPECS=1 and CYTHON_USE_TYPE_SLOTS=0
### Describe your issue

Working through the failures in PyPy when `-DCYTHON_USE_TYPE_SPECS=1` (after #5594), I am [down to one](https://github.com/pypy/binary-testing/actions/runs/6155039760/job/16701328208#step:7:12901): `cdef_multiple_inheritance_nodict`. It hits this error:
> `unable to validate whether bases have a __dict__ when CYTHON_USE_TYPE_SLOTS is off (likely because you are building in the limited API). Therefore, all extension types with multiple bases must add 'cdef dict __dict__' in this compilation mode`

from [here](https://github.com/cython/cython/blob/ce1aa59ab6300e18e50c752b5c9e28a04b665060/Cython/Utility/ExtensionTypes.c#L158). Am I correct that the failure is expected, and so the test should expect to hit that message when testing this configuration?

### PR DESCRIPTION
**Pull request by da-woods**
Rather than just declaring it impossible.

Fixes #5696

(Note that this isn't sufficient to declare multiply-inherited cdef classes in the limited api - the calling code generates a few `PyTuple_GET_ITEM`s. However, I have tested the dictoffset code itself)

### CODE REVIEW

File: Cython/Utility/ExtensionTypes.c [hunk 1]
```diff
@@ -155,38 +155,36 @@ static int __Pyx_validate_bases_tuple(const char *type_name, Py_ssize_t dictoffs
...
+            b =  = b->tp_dictoffset;
+#else
+            PyObject *py_b_dictoffset = PyObject_GetAttrString((PyObject*)b, "__dictoffset__");
+            if (!py_b_dictoffset) goto dictoffset_return;
+            b_dictoffset = PyLong_AsSsize_t(py_b_dictoffset);
+            Py_DECREF(py_b_dictoffset);
+            if (b_dictoffset == -1 && PyErr_Occurred()) goto dictoffset_return;
+#endif
+            if (b_dictoffset) {
+                __Pyx_TypeName b_name = __Pyx_PyType_GetName(b);
```
Comment (mattip) [inline Cython/Utility/ExtensionTypes.c:171]: The declaration of `b_name` is between the `goto` and the label, which will fail in C++

Comment (mattip) [reply Cython/Utility/ExtensionTypes.c:171]: Adding another set of braces to enclose the scope of `b_name` fixes the c++ compilation error, but then running the test on PyPy I get
```
CFLAGS="-DCYTHON_USE_TYPE_SPECS=1" ../pypy/pypy3.9-c runtests.py cdef_multiple_inheritance_nodict --no-c --no-cleanup
...
File "tests/run/cdef_multiple_inheritance_nodict.pyx", line 16, in init cdef_multiple_inheritance_nodict (cdef_multiple_inheritance_nodict.cpp:7472)
    cdef class Both(CBase, PyBase):
AttributeError: type object 'PyBase' has no attribute '__dictoffset__'
```

which seems to be an incompatibility in PyPy:  classes never have `__dictoffset__`. It turns out there is an [open issue](https://foss.heptapod.net/pypy/pypy/-/issues/2447) about the lack of proper handling of `_PyObject_GetDictPtr` and `tp_dictoffset` for pybind11, that was never resolved. So while this PR fixes a corner case for CPython, it doesn't fix the PyPy test failure.

Sorry for the bother. I am not sure how

Comment (mattip) [reply Cython/Utility/ExtensionTypes.c:171]: well, I guess I could get `USE_CYTHON_TYPE_SLOTS` to work on PyPy.

File: Cython/Utility/ExtensionTypes.c [hunk 2]
```diff
@@ -155,38 +155,36 @@ static int __Pyx_validate_bases_tuple(const char *type_name, Py_ssize_t dictoffs
...
-                "or add '__slots__ = [...]' to the base type",
-                type_name, b_name);
-            __Pyx_DECREF_TypeName(b_name);
-#if CYTHON_AVOID_BORROWED_REFS
-            Py_DECREF(b0);
-#endif
-            return -1;
+            Py_ssize_t b_dictoffset = 0;
+#if CYTHON_USE_TYPE_SLOTS
+            b =  = b->tp_dictoffset;
```
Comment (da-woods) [inline Cython/Utility/ExtensionTypes.c:162]:
```suggestion
            b = b->tp_dictoffset;
```

File: Cython/Utility/ExtensionTypes.c [hunk 3]
```diff
@@ -155,38 +155,36 @@ static int __Pyx_validate_bases_tuple(const char *type_name, Py_ssize_t dictoffs
...
+                    "or add '__slots__ = [...]' to the base type",
+                    type_name, b_name);
+                __Pyx_DECREF_TypeName(b_name);
+                {
+                  dictoffset_return:
+    #if CYTHON_AVOID_BORROWED_REFS
+                    Py_DECREF(b0);
+    #endif
+                    return -1;
+                }
```
Comment (da-woods) [inline Cython/Utility/ExtensionTypes.c:185]:
```suggestion
            if (b_dictoffset) {
                {
                    __Pyx_TypeName b_name = __Pyx_PyType_GetName(b);
                    PyErr_Format(PyExc_TypeError,
                        "extension type '%.200s' has no __dict__ slot, "
                        "but base type '" __Pyx_FMT_TYPENAME "' has: "
                        "either add 'cdef dict __dict__' to the extension type "
                        "or add '__slots__ = [...]' to the base type",
                        type_name, b_name);
                    __Pyx_DECREF_TypeName(b_name);
                }
    #if CYTHON_USE_TYPE_SLOTS
              dictoffset_return:
    #endif
    #if CYTHON_AVOID_BORROWED_REFS
                Py_DECREF(b0);
    #endif
                return -1;
```

File: Cython/Utility/ExtensionTypes.c [hunk 4]
```diff
@@ -155,38 +155,38 @@ static int __Pyx_validate_bases_tuple(const char *type_name, Py_ssize_t dictoffs
...
-                "either add 'cdef dict __dict__' to the extension type "
-                "or add '__slots__ = [...]' to the base type",
-                type_name, b_name);
-            __Pyx_DECREF_TypeName(b_name);
-#if CYTHON_AVOID_BORROWED_REFS
-            Py_DECREF(b0);
-#endif
-            return -1;
+            Py_ssize_t b_dictoffset = 0;
+#if CYTHON_USE_TYPE_SLOTS
```
Comment (da-woods) [inline Cython/Utility/ExtensionTypes.c:161]:
```suggestion
#if CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY
```

File: Cython/Utility/ExtensionTypes.c (hunk 1 above)
Comment (da-woods) [reply Cython/Utility/ExtensionTypes.c]: I made the test

```
#if CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY
```

Not completely sure if this'll work or not though, but it seems easier than trying to get the *whole* of CYTHON_USE_TYPE_SLOTS to work with PyPy.

> Sorry for the bother. I am not sure how to fix this.

It's still worth fixing the limited api side of it anyway, even if it doesn't fix PyPy.

File: Cython/Utility/ExtensionTypes.c [hunk 5]
```diff
@@ -155,38 +155,38 @@ static int __Pyx_validate_bases_tuple(const char *type_name, Py_ssize_t dictoffs
...
-                "or add '__slots__ = [...]' to the base type",
-                type_name, b_name);
-            __Pyx_DECREF_TypeName(b_name);
-#if CYTHON_AVOID_BORROWED_REFS
-            Py_DECREF(b0);
-#endif
-            return -1;
+            Py_ssize_t b_dictoffset = 0;
+#if CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY
+            b = b->tp_dictoffset;
```
Comment (mattip) [inline Cython/Utility/ExtensionTypes.c:162]:
```suggestion
            b_dictoffset = b->tp_dictoffset;
```

File: Cython/Utility/ExtensionTypes.c [hunk 6]
```diff
@@ -155,38 +155,38 @@ static int __Pyx_validate_bases_tuple(const char *type_name, Py_ssize_t dictoffs
...
+                    __Pyx_TypeName b_name = __Pyx_PyType_GetName(b);
+                    PyErr_Format(PyExc_TypeError,
+                        "extension type '%.200s' has no __dict__ slot, "
+                        "but base type '" __Pyx_FMT_TYPENAME "' has: "
+                        "either add 'cdef dict __dict__' to the extension type "
+                        "or add '__slots__ = [...]' to the base type",
+                        type_name, b_name);
+                    __Pyx_DECREF_TypeName(b_name);
+                }
+    #if CYTHON_USE_TYPE_SLOTS
```
Comment (da-woods) [inline Cython/Utility/ExtensionTypes.c:181]:
```suggestion
    #if !CYTHON_USE_TYPE_SLOTS
```

File: Cython/Utility/ExtensionTypes.c [hunk 7]
```diff
@@ -155,38 +155,38 @@ static int __Pyx_validate_bases_tuple(const char *type_name, Py_ssize_t dictoffs
...
+                    __Pyx_TypeName b_name = __Pyx_PyType_GetName(b);
+                    PyErr_Format(PyExc_TypeError,
+                        "extension type '%.200s' has no __dict__ slot, "
+                        "but base type '" __Pyx_FMT_TYPENAME "' has: "
+                        "either add 'cdef dict __dict__' to the extension type "
+                        "or add '__slots__ = [...]' to the base type",
+                        type_name, b_name);
+                    __Pyx_DECREF_TypeName(b_name);
+                }
+    #if !CYTHON_USE_TYPE_SLOTS
```
Comment (scoder) [inline Cython/Utility/ExtensionTypes.c:181]: You use a different indentation for the `#if` here than above.
Also, this should probably read
```suggestion
#if !(CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY)
```
to match the condition above.

**Review by scoder — APPROVED** [review]
Good improvement, just one thing to fix.

**Comment by da-woods** [comment]
Seems to have broken something in PyPy.

I'm away from my PC for the rest of the week so I'll investigate properly when I get back.

**Comment by scoder** [comment]
> Seems to have broken something in PyPy.

No, it's fine, that was a hickup in master on my part.
